Content.Server/Radiation/Systems/RadiationSystem.GridCast.cs

```diff
@@ -0,0 +1,234 @@
+    {
+        var visitedTiles = new List<(Vector2i, float?)>();
+        var radRay = new RadiationRay(mapId, sourceUid,sourceWorld,
+            destUid,destWorld, incomingRads)
+        {
+            Grid = gridUid,
+            VisitedTiles = visitedTiles
+        };
+
+        // if grid doesn't have resistance map just apply distance penalty
+        if (!_resistancePerTile.TryGetValue(gridUid, out var resistanceMap))
+            return radRay;
+
+        // get coordinate of source and destination in grid coordinates
+        // todo: entity queries doesn't support interface - use it when IMapGridComponent will be removed
+        if (!TryComp(gridUid, out IMapGridComponent? grid))
+            return radRay;
+        var sourceGrid = grid.Grid.TileIndicesFor(sourceWorld);
+        var destGrid = grid.Grid.TileIndicesFor(destWorld);
+
+        // iterate tiles in grid line from source to destination
+        var line = Line(sourceGrid.X, sourceGrid.Y, destGrid.X, destGrid.Y);
+        foreach (var point in line)
+        {
+            if (resistanceMap.TryGetValue(point, out var resData))
+            {
+                radRay.Rads -= resData;
+                visitedTiles.Add((point, radRay.Rads));
+            }
+            else
+            {
+                visitedTiles.Add((point, null));
+            }
+
+            // no intensity left after blocker
+            if (radRay.Rads <= MinIntensity)
+            {
+                radRay.Rads = 0;
+                return radRay;
+            }
+        }
+
+        return radRay;
+    }
+
+    private RadiationRay Raycast(MapId mapId, EntityUid sourceUid, EntityUid destUid,
+        Vector2 sourceWorld, Vector2 destWorld, Vector2 dir, float distance, float incomingRads,
+        EntityQuery<RadiationBlockerComponent> blockerQuery)
+    {
+        var blockers = new List<(Vector2, float)>();
+        var radRay = new RadiationRay(mapId, sourceUid, sourceWorld,
+            destUid, destWorld, incomingRads)
+        {
+            Blockers = blockers
+        };
+
+        var colRay = new CollisionRay(sourceWorld, dir, (int) CollisionGroup.Impassable);
+        var results = _physicsSystem.IntersectRay(mapId, colRay, distance, returnOnFirstHit: false);
+
+        foreach (var obstacle in results)
+        {
+            if (!blockerQuery.TryGetComponent(obstacle.HitEntity, out var blocker))
+                continue;
+
+            radRay.Rads -= blocker.RadResistance;
+            blockers.Add((obstacle.HitPos, radRay.Rads));
+
+            if (radRay.Rads <= MinIntensity)
+            {
+                radRay.Rads = 0;
+                return radRay;
+            }
+        }
+
+        return radRay;
+    }
+
+
+    // bresenhams line algorithm
+    // this is slightly rewritten version of code bellow
+    // https://stackoverflow.com/questions/11678693/all-cases-covered-bresenhams-line-algorithm
+    private IEnumerable<Vector2i> Line(int x, int y, int x2, int y2)
+    {
```

Comment:
1. Bresenham's cuts diagonals which is unwanted behaviour
2. If you care about performance don't use IEnumerables.

Reply:
> If you care about performance don't use IEnumerables.

The most straightforward way would be to embed cycle body inside loop. But code becomes so unreadable...

Other way is to do some micro optimizations, like used typed enumerator, but I agree with PJB that they probably not needed here.

Reply:
> Bresenham's cuts diagonals which is unwanted behaviour

While cutting diagonals is really undesirable for things like atmos or explosions, IMO its fine for radiation. Not only is it invisible and sort of expected to go through walls, but it is also true that cutting across the corner of a wall should block much less radiation than cutting along a diagonal, so including the whole wall as a radiation blocker is arguably less realistic.

Reply:
> I guess you could also just add some algorithm that includes every touching grid tile, along with the length of the line that actually intersected each tile, which could be used to scale the degree of blocking. But that would also just be slower and probably not worth it.

It will be something like [antialiasing of Wu algorithm](https://en.wikipedia.org/wiki/Xiaolin_Wu%27s_line_algorithm), but this really overengineering.